mtcpclient/groupofconnections.go

```diff
@@ -29,23 +48,89 @@ func (gc GroupOfConnections) String() string {
 		nTotal, nDialing, nEstablished, nClosed, nError, nNotInitiated)
 }
 
-func (gc GroupOfConnections) isIn(status tcpclient.ConnectionStatus) bool {
-	for _, item := range gc {
+func (gc GroupOfConnections) containsAConnectionWithStatus(status tcpclient.ConnectionStatus) bool {
+	for _, item := range gc.connections {
 		if item.GetConnectionStatus() == status {
 			return true
 		}
 	}
 	return false
 }
 
+// PendingConnections retuns True if at least one connection is being processed
 func (gc GroupOfConnections) PendingConnections() bool {
-	return gc.isIn(tcpclient.ConnectionNotInitiated) || gc.isIn(tcpclient.ConnectionDialing)
+	for _, connection := range gc.connections {
+		if connection.PendingToProcess() {
+			return true
+		}
+	}
+	return false
 }
 
+// AtLeastOneConnectionInError returns True is at least one connection establishment failed
 func (gc GroupOfConnections) AtLeastOneConnectionInError() bool {
-	return gc.isIn(tcpclient.ConnectionError)
+	for _, connection := range gc.connections {
+		if connection.WithError() {
+			return true
+		}
+	}
+	return false
+}
+
+func (gc GroupOfConnections) atLeastOneConnectionOK() bool {
+	for _, connection := range gc.connections {
+		if connection.WentOk() {
+			return true
+		}
+	}
+	return false
+}
+
+func (gc GroupOfConnections) pingStyleReport(typeOfReport string) (output string) {
+	var connectionsFiltered GroupOfConnections
+	switch typeOfReport {
+	case "successful":
+		connectionsFiltered = gc.getConnectionsThatWentWell()
```

Comment:
I'm afraid i think we can still improve this function. A function that starts by a "switch" and executes two different code sets depending on a condition is a very bad sign. See what these two blocks have in common, and consolidate it. Something that may be helpful is to move some staff to the caller; he can do something like: gc.connectionsThatWent(well()).pingStyleReport("success")

Reply:
done